--- mialib_wrap.c	2016-09-21 14:55:24.946731224 +0200
+++ mialib_wrap.c.PATCH.2016092201	2016-09-21 15:09:09.278818866 +0200
@@ -1548,6 +1548,9 @@
   return SwigPyObject_format("%x",v);
 }
 
+/* patch by Pierre.Soille@jrc.ec.europa.eu */
+#include "mialib_swig.h"
+
 SWIGRUNTIME PyObject *
 #ifdef METH_NOARGS
 SwigPyObject_repr(SwigPyObject *v)
@@ -1557,6 +1560,26 @@
 {
   const char *name = SWIG_TypePrettyName(v->ty);
   PyObject *repr = SWIG_Python_str_FromFormat("<Swig Object of type '%s' at %p>", (name ? name : "unknown"), (void *)v);
+  /* start of patch 3 by Pierre.Soille@jrc.ec.europa.eu */
+#ifdef _DEBUG
+  printf("\n bonjour in entering SwigPyObject_repr\n");
+  printf("coucou in SwigPyObject_repr()\n");
+#endif
+  if (strcmp(v->ty->str, "IMAGE *")==0){
+#ifdef _DEBUG
+    printf("\t v->own=%d\n", v->own);
+    printf("\t v->next=%p\n", v->next);
+    printf("\t v object reference count of v = %d\n", (int)Py_REFCNT((PyObject *)v) );
+    printf("\t v object reference count of repr = %d\n", (int)Py_REFCNT((PyObject *)repr) );
+#endif
+    if (v->own == 0 && Py_REFCNT((PyObject *)v)==22 ){ /* 22 determined experimentally ... */
+#ifdef _DEBUG
+      printf("patch3 to avoid memory leak with zombie image\n");
+#endif
+      free_image(v->ptr);
+    }
+  }
+  /* end of patch 3*/
   if (v->next) {
 # ifdef METH_NOARGS
     PyObject *nrep = SwigPyObject_repr((SwigPyObject *)v->next);
@@ -1661,10 +1684,31 @@
 #if !defined(SWIG_PYTHON_SILENT_MEMLEAK)
     else {
       const char *name = SWIG_TypePrettyName(ty);
+#ifdef _DEBUG
       printf("swig/python detected a memory leak of type '%s', no destructor found.\n", (name ? name : "unknown"));
+      /* start of patch 1 by Pierre.Soille@jrc.ec.europa.eu */
+      printf("patch1 %s\n", name);
+#endif
+      if (strcmp(name, "IMAGE *")==0){
+#ifdef _DEBUG
+	printf("patch1 to avoid memory leak with image with sobj->own=1\n");
+#endif
+	free_image(sobj->ptr);	 
+      }
+      /* end of patch 1 by Pierre.Soille@jrc.ec.europa.eu */
     }
 #endif
-  } 
+  }   
+  /* start of patch 2 by Pierre */
+  else if (strcmp(sobj->ty->str, "IMAGE *")==0){
+#ifdef _DEBUG
+    printf("in patch2 %s\n", sobj->ty->str);
+    printf("patch2 to avoid memory leak with image with sobj->own=0\n");
+#endif
+    free_image(sobj->ptr);
+  }
+  /* end of patch 2 by Pierre.Soille@jrc.ec.europa.eu */
+  
   Py_XDECREF(next);
   PyObject_DEL(v);
 }
